src/centralstore/incremental/NodeManager.cpp

```diff
@@ -0,0 +1,254 @@
+limitations under the License.
+**/
+
+#include "NodeManager.h"
+
+#include <sys/stat.h>
+
+#include "../../util/logger/Logger.h"
+#include "NodeBlock.h"  // To setup node DB
+#include "PropertyLink.h"
+#include "RelationBlock.h"
+
+Logger node_manager_logger;
+
+NodeManager::NodeManager(std::string mode) {
+    std::ios_base::openmode openMode = std::ios::trunc;  // default is Trunc mode which overrides the entire file
+    if (mode == "app") {
+        openMode = std::ios::app;  // if app, open in append mode
+        this->nodeIndex = readNodeIndex();
+    }
+    NodeBlock::nodesDB =
+        new std::fstream(NodeManager::NODE_DB_PATH, std::ios::in | std::ios::out | openMode | std::ios::binary);
+    PropertyLink::propertiesDB =
+        new std::fstream(PropertyLink::DB_PATH, std::ios::in | std::ios::out | openMode | std::ios::binary);
+    RelationBlock::relationsDB =
+        new std::fstream(RelationBlock::DB_PATH, std::ios::in | std::ios::out | openMode | std::ios::binary);
+    // TODO (tmkasun): set PropertyLink nextPropertyIndex after validating by modulus check from file number of bytes
+
+    if (dbSize(NodeManager::NODE_DB_PATH) % NodeBlock::BLOCK_SIZE != 0) {
+        node_manager_logger.error("Node DB size does not comply to node block size Path = " +
+                                  NodeManager::NODE_DB_PATH);
+    }
+}
+
+std::unordered_map<std::string, unsigned int> NodeManager::readNodeIndex() {
+    std::ifstream index_db(this->index_db_loc, std::ios::app | std::ios::binary);
+    std::unordered_map<std::string, unsigned int> _nodeIndex;  // temporary node index data holder
+
+    if (index_db.is_open()) {
+        int iSize = dbSize(this->index_db_loc);
+        unsigned long dataWidth = NodeManager::INDEX_KEY_SIZE + sizeof(unsigned int);
+        if (iSize % dataWidth != 0) {
+            node_manager_logger.error("Index DB size does not comply to index block size Path = " +
+                                      this->index_db_loc);
+            throw std::runtime_error("Node index DB in " + this->index_db_loc + " is corrupted!");
+        }
+
+        char nodeIDC[NodeManager::INDEX_KEY_SIZE];
+        unsigned int nodeIndexId;
+        for (size_t i = 0; i < iSize / dataWidth; i++) {
+            nodeIDC[NodeManager::INDEX_KEY_SIZE] = {0};  // Fill with null chars before puting data
+            if (!index_db.read(&nodeIDC[0], NodeManager::INDEX_KEY_SIZE)) {
+                node_manager_logger.error("Error while reading index key data from block i = " + std::to_string(i));
+            }
+            if (!index_db.read(reinterpret_cast<char *>(&nodeIndexId), sizeof(unsigned int))) {
+                node_manager_logger.error("Error while reading index ID data from block i = " + std::to_string(i));
+            }
+            _nodeIndex[std::string(nodeIDC)] = nodeIndexId;
+        }
+    } else {
+        node_manager_logger.error("Error while opening the node index DB");
```

Comment:
why don't we throw an error here?

Reply:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

Reply:
Done